File: ruby/numerals_test.rb
require "./numerals.rb"
require "minitest/autorun"

class NumeralsTest < Minitest::Test

  def test_parses_number_by_thousand_pow_to_term
    assert_equal [[0, 0, 7],
                  [6, 5, 4],
                  [3, 2, 1]],
                 number_to_term(7_654_321)

    assert_equal [[0, 3, 7],
                  [6, 5, 4],
                  [3, 2, 1]],
                 number_to_term(37_654_321)

    assert_equal [[0, 0, 5],
                  [4, 3, 7],
                  [6, 5, 4],
                  [3, 2, 1]],
                 number_to_term(5_437_654_321)

    assert_equal [[0, 0, 5],
                  [4, 3, 7],
                  [0, 0, 0],
                  [3, 2, 1]],
                 number_to_term(5_437_000_321)

    assert_equal [[0, 0, 1],
                  [0, 0, 0]],
                 number_to_term(1_000)
  end

  def test_transform_ones_to_text
    assert_equal "one",   one_to_text(1)
    assert_equal "two",   one_to_text(2)
    assert_equal "three", one_to_text(3)
    assert_equal "four",  one_to_text(4)
    assert_equal "five",  one_to_text(5)
    assert_equal "six",   one_to_text(6)
    assert_equal "seven", one_to_text(7)
    assert_equal "eight", one_to_text(8)
    assert_equal "nine",  one_to_text(9)
    assert_equal "zero",  one_to_text(0)
  end

  def test_transform_tens_to_text
    assert_equal "twenty-two",   ten_to_text([2, 2])
    assert_equal "thirty-three", ten_to_text([3, 3])
    assert_equal "ten",          ten_to_text([1, 0])
    assert_equal "eleven",       ten_to_text([1, 1])
    assert_equal "eighteen",     ten_to_text([1, 8])
    assert_equal "sixteen",      ten_to_text([1, 6])
  end


  def test_transform_hundreds_to_text
    assert_equal "one hundred",            hundred_to_text([1, 0, 0])
    assert_equal "six hundred two",        hundred_to_text([6, 0, 2])
    assert_equal "one hundred nineteen",   hundred_to_text([1, 1, 9])
    assert_equal "one hundred twenty",     hundred_to_text([1, 2, 0])
    assert_equal "two hundred twenty-two", hundred_to_text([2, 2, 2])
  end

  def test_transform_hundreds_with_pow_to_text
    assert_equal "one",          hundred_with_pow_to_text([0, 0, 1], 0)
    assert_equal "one thousand", hundred_with_pow_to_text([0, 0, 1], 1)
    assert_equal "one million",  hundred_with_pow_to_text([0, 0, 1], 2)
    assert_equal "one billion",  hundred_with_pow_to_text([0, 0, 1], 3)
  end

  def test_transform_term_to_text
    assert_equal "seven million, six hundred fifty-four thousand, three hundred twenty-one",
                 term_to_text([[0, 0, 7],
                               [6, 5, 4],
                               [3, 2, 1]])
    assert_equal "five billion, four hundred thirty-seven million, three hundred twenty-one",
                 term_to_text([[0, 0, 5],
                               [4, 3, 7],
                               [0, 0, 0],
                               [3, 2, 1]])
  end

  def test_to_numeral
    assert_equal "seven million, six hundred fifty-four thousand, three hundred twenty-one",
                 to_numeral(7_654_321)
    assert_equal "five billion, four hundred thirty-seven million, six hundred fifty-four thousand, three hundred twenty-one",
                 to_numeral(5_437_654_321)
    assert_equal "zero", to_numeral(0)
  end

end

File: javascript/test/numerals_test.js
const assert = require('assert')
const numerals = require('./../numerals.js')

describe('Numerals', function () {
  describe('#numberToTermshould', function () {
    it('should return list of hundred nodes', function () {
      assert.deepEqual([[0, 0, 7],
                        [6, 5, 4],
                        [3, 2, 1]],
                       numerals.numberToTerm(7654321))
      assert.deepEqual([[0, 3, 7],
                        [6, 5, 4],
                        [3, 2, 1]],
                       numerals.numberToTerm(37654321))
      assert.deepEqual([[0, 0, 5],
                        [4, 3, 7],
                        [6, 5, 4],
                        [3, 2, 1]],
                       numerals.numberToTerm(5437654321))
      assert.deepEqual([[0, 0, 5],
                        [4, 3, 7],
                        [0, 0, 0],
                        [3, 2, 1]],
                       numerals.numberToTerm(5437000321))
      assert.deepEqual([[0, 0, 1],
                        [0, 0, 0]],
                       numerals.numberToTerm(1000))
    })
  })

  describe('#oneToText', function () {
    it('transform ones to text', function () {
      assert.equal('one',   numerals.oneToText(1))
      assert.equal('two',   numerals.oneToText(2))
      assert.equal('three', numerals.oneToText(3))
      assert.equal('four',  numerals.oneToText(4))
      assert.equal('five',  numerals.oneToText(5))
      assert.equal('six',   numerals.oneToText(6))
      assert.equal('seven', numerals.oneToText(7))
      assert.equal('eight', numerals.oneToText(8))
      assert.equal('nine',  numerals.oneToText(9))
    })
  })

  describe('#tenToText', function () {
    it('transform tens to text', function () {
      assert.equal('twenty-two',   numerals.tenToText([2, 2]))
      assert.equal('thirty-three', numerals.tenToText([3, 3]))
      assert.equal('ten',          numerals.tenToText([1, 0]))
      assert.equal('eleven',       numerals.tenToText([1, 1]))
      assert.equal('eighteen',     numerals.tenToText([1, 8]))
      assert.equal('sixteen',      numerals.tenToText([1, 6]))
    })
  })

  describe('#hundredToText', function () {
    it('transform hundreds to text', function () {
      assert.equal('one hundred',            numerals.hundredToText([1, 0, 0]))
      assert.equal('six hundred two',        numerals.hundredToText([6, 0, 2]))
      assert.equal('one hundred nineteen',   numerals.hundredToText([1, 1, 9]))
      assert.equal('one hundred twenty',     numerals.hundredToText([1, 2, 0]))
      assert.equal('two hundred twenty-two', numerals.hundredToText([2, 2, 2]))
    })
  })

  describe('#hundredWithPowToText', function () {
    it('transform hundreds with pow to text', function () {
      assert.equal("one",          numerals.hundredWithPowToText([0, 0, 1], 0))
      assert.equal("one thousand", numerals.hundredWithPowToText([0, 0, 1], 1))
      assert.equal("one million",  numerals.hundredWithPowToText([0, 0, 1], 2))
      assert.equal("one billion",  numerals.hundredWithPowToText([0, 0, 1], 3))
    })
  })

  describe('#termToText', function () {
    it('transform term to text', function () {
      assert.equal('seven million, six hundred fifty-four thousand, three hundred twenty-one',
                   numerals.termToText([[0, 0, 7],
                                        [6, 5, 4],
                                        [3, 2, 1]]))
      assert.equal('five billion, four hundred thirty-seven million, three hundred twenty-one',
                   numerals.termToText([[0, 0, 5],
                                        [4, 3, 7],
                                        [0, 0, 0],
                                        [3, 2, 1]]))
    })
  })

  describe('#toNumeral', function () {
    it('transform number to english numeral', function () {
      assert.equal('seven million, six hundred fifty-four thousand, three hundred twenty-one',
                   numerals.toNumeral(7654321))
      assert.equal('five billion, four hundred thirty-seven million, six hundred fifty-four thousand, three hundred twenty-one',
                   numerals.toNumeral(5437654321))
      assert.equal('zero', numerals.toNumeral(0))
    })
  })
})
